pkg/base/backoff/handler.go

```diff
@@ -0,0 +1,48 @@
+package backoff
+
+import (
+	"errors"
+	"math/rand"
+	"time"
+)
+
+// ErrMaxBackoffRetriesExceeded stays for max backoff retries exceeded error
+var ErrMaxBackoffRetriesExceeded = errors.New("max backoff retries exceeded")
+
+// Handler is a backoff handler
+type Handler interface {
+	Handle(retries int) time.Duration
+}
+
+type exponentialHandler struct {
+	config *Config
+}
+
+func (s *exponentialHandler) Handle(retries int) time.Duration {
+	// if there were no retries, then return baseDelay time, which is 1 second by default
+	if retries == 0 {
+		return s.config.BaseDelay
+	}
+	// recalculating current backoff and comparing with maximum
+	curBackoff, max := float64(s.config.BaseDelay), float64(s.config.MaxDelay)
+	for curBackoff < max && retries > 0 {
+		curBackoff *= s.config.Multiplier
+		retries--
+	}
+	if curBackoff > max {
+		curBackoff = max
+	}
+	// randomize backoff delays so that if cluster of requests start at
+	// the same time, they won't operate in lockstep.
+	if curBackoff *= 1 + s.config.Jitter*(rand.Float64()*2-1); curBackoff < 0 {
+		return 0
+	}
+	return time.Duration(curBackoff)
+}
+
+// NewExponentialHandler is a constructor for backoff handler
+func NewExponentialHandler(config *Config) Handler {
```

Comment:
What do you think about a little changes I suggest here? 

```golang
// Handler is a backoff handler
type Handler interface {
	ExponentialHandle(retries int) time.Duration
}

type handler struct {
	config *Config
}
func (s *handler) ExponentialHandle(retries int) time.Duration {
	// some code here
}

func NewHandler(config *Config) Handler {
       return &handler{
		config: config,
	}
}
```

This code is much readable and expandable because only `ExponentialHandle` method doing exponential backoff, so we can introduce another method with for example `LinearHandle` with a little changes just in package.

Reply:
We could do so, but in this case we should pass special configurations for each type of backoff (e.g. pass jitter to exponential only) and it would be excessive for caller methods to handle this information due to the nature of backoff method - it should be peripheral and easy to use.
Moreover by allowing to pass this configuration parameters for each run caller has ability to manage exponentiality of backoff - possible making it to not become exponential at all.
For now, as for me, it's better to remain logic as it is.

Reply:
You don't need special configuration, just extend default config. Dont see a problem with configuration could you show any example?